### DISCUSSION TITLE
Ditching the sidebar

### DISCUSSION THREAD

**Discussion by oreilles** [directus/directus#3685] [Feature Requests]
I discovered directus with the latest version and I find it to be absolutely amazing.
However I can't stand the (right) sidebar:
1. It clutters the UI, taking a lot of space on the right, very noticable when you have even just more than a few rows in your table. Sure you can fold it, but then you lose direct access to all of its features, leading to n°2:
2. Every sidebar field (layout mode, layout options, filtering) should be accessible  easily and without friction from the main view.  It takes 4 click instead of 2 to interact with your data or the layout, in addition to the fact that you have to readjust your eyes to the layout width.
3. There is no clear benefit in hiding those options in a side panel. It is just as restrictive size-wise as it if were in the header bar with the search field.

And that's when I realized the previous version had it all right:

<img width="862" alt="Capture d’écran 2021-01-14 à 00 41 08" src="https://user-images.githubusercontent.com/33065839/104525954-9b932580-5601-11eb-9ca9-2a5ea898a6e0.png">

I don't understand that design change. Instead of pushing feature out of the header bar and on to the sidebar, I think we should to the opposite, ditching the sidebar completely and pushing everything to the header bar, cleary visible and accessible. It's how almost any data visualization tool I know does it and it's the way to go 🙂

**Comment by benhaynes** [comment]
Thanks @Oreilles!! I love really diving into the UX/UI optimizations. Obviously I spent a lot of time thinking through how I thought this would be best, so let me give you my thoughts on that, then we can keep discussing...

* **Real-Estate**: The header was getting a bit crowded horizontally, and we needed a place to put secondary/tertiary features. This was especially important for responsive, since we have very little space in the header on narrow devices. Also, this sidebar offers more space _vertically_, so you can see more info without it covering your page.
* **Extensibility**: Similar to the space issue, but also adding a "modular" UI, we needed a place where custom code modules could be places on the page in a logical way. The header is seen as "sacred" because it's so important to good user flow... so I was reluctant to open that up for customization (as of now).
* **Accessibility**: By this I mean click intuitiveness (finding what you're looking for) and click-count to access. 
    * Intuitiveness: Yes, advanced filtering went well with search... so it had a natural "home", but it would be hard to also find places for Layout Options, Import/Export, Page Info, etc... not to mention any of the Custom Modules users may choose to add.
    * Click-Count: For Advanced Search, in the v8 header you had three clicks to add a filter: click search (expands), click advanced search (opens popup), click "Add Filter". In v9, you can click the Advanced Search module in the sidebar, then "Add Filter"... two clicks. 
    * Visibility: One advantage of the sidebar is that you can have it open and still work. So you can actually see your filters while working with the data. This is especially important for other Layouts that may need more advanced customization.
    * Global: Headers can be different on each page, and some pages might want a very minimized header (eg:  Map Layout)... without the sidebar, all of our current and future features would need to find SOME way of fitting up there... making it bigger.
* **Size**: It automatically collapses when screen size is smaller, so it shouldn't be in the way too much. I think one solution to the data reflow would be to allow pages to decide if the sidebar should "compress" the page, or "slide over it" (this is how it works on mobile). This would avoid things shifting around on pages where that is not desired.
* **Changes**: _Some_ features could still move from the sidebar to the header. For instance, I could imagine sorting/toggling Table Layout columns could be seamlessly integrated into the table header. But this would need more thinking.

The most important thing to understand from all this is that the sidebar is meant to allow for extensibility... though this is not available yet. This modular/decoupled approach was the best way that I could find to give those customizations room, and not let them "screw up" the Directus UI.

Interested to hear your thoughts on this!

**Reply by philleepflorence** [reply]
I noticed the **Filter** in the Side Bar and it felt intuitive, I had to let clients know to click the **Filter** Icon in D8 to filter results.
Extending the Side Bar is going to bring a lot of smiles on my end!

**Comment by oreilles** [comment]
Thanks for your insight! 

> Also, this sidebar offers more space vertically, so you can see more info without it covering your page.

But the space it offers vertically is only usable when expanded, therefore _also_ taking space horizontally. You also don't need to see informations about your data at the same time you browse your data, anyway doing both at the same time is not feasible - it's even less feasible that the left and right sidebar are talking so much space when opened that I can barely show 4 rows on my laptop, not good for checking my data at the same time even if I wanted to.

>  Accessibility Intuitiveness: Yes, advanced filtering went well with search... so it had a natural "home", but it would be hard to also find places for Layout Options, Import/Export, Page Info, etc... not to mention any of the Custom Modules users may choose to add.

Layout Options, Import/Export, should all go to the header bar. That's just 2 icons to add (if the filters one goes with the search icon as in v8), they would very easily fit into the header, especially considering that on small screen the search bar is anyway already hidden bellow an epandable. Same could go with the other icons .

> Click-Count: For Advanced Search, in the v8 header you had three clicks to add a filter: click search (expands), click advanced search (opens popup), click "Add Filter". In v9, you can click the Advanced Search module in the sidebar, then "Add Filter"... two clicks.

If the v8 in the one at [demo.directus.io](https://demo.directus.io/) you actually only need two clicks, 1 on the filter icon, 1 on "Add filter". So it's at best the same click count, but that's only assuming your sidebar is open, which probably isn't the case if you have more than a few rows. Most of the time it will require 3 click. To give context, on my screen the left panel, sidebar and scrollbar are taking **40%** of the screen real estate 😨 (264px + 284px + 12px) / 1400. That really isn't good condition to browse data.

> Visibility: One advantage of the sidebar is that you can have it open and still work. So you can actually see your filters while working with the data. This is especially important for other Layouts that may need more advanced customization.

That's still assuming the sidebar is open (I'm not going to repeat why it wouldn't be 😊)

In order to give weight to my points and show you that you can have the best of every world (rich and customizable layouts, quick feature discovering and access, uncluttered interface) I'm going to show how Notion does it. I know it's the second time I bring it up but I think it does so many things right 😁 And also because Directus has a lot features in common and I think it could really be _the_ open source, self hosted, highly customizable alternative to Notion, which doesn't exist yet, with only few design improvements.
<p align="top">
<img width="400" src="https://user-images.githubusercontent.com/33065839/104620788-aac0b480-568f-11eb-8376-a26be15bdf62.png">
  <img  width="400" src="https://user-images.githubusercontent.com/33065839/104622090-1d7e5f80-5691-11eb-973d-c97cba6313de.png">
</p>

Layout mode change (`All tasks`), Layout options (`Properties`), advanced filtering, simple search, new item creation, and much more (including csv export) is visible and only one click away, in only 630 pixel and even with some space left ! Desciption about the collection is hidden if empty, visible by default if not but can be hidden with a click, Title and descriptions will be hidden on scroll leaving a clean view of your data, table header staying sticky to the top.

Item count is sticky to the bottom. It's also also a great position as it unclutters the header and would fit well with the pages numbers / limit options - that would also then be available without scrolling all the way down 😃.

**Reply by benhaynes** [reply]
Some good points, thanks. Also, I love Notion and I'm sure there are good UX/UI things they've done that we could learn from. Some quick responses below...

> ...but that's only assuming your sidebar is open, which probably isn't the case if you have more than a few rows. Most of the time it will require 3 click

If you click on the sidebar Advanced Filter icon (when collapsed) it opens that component directly. No need to toggle the sidebar open first.

Lastly, a lot of your points are based on your specific viewport size, presumably a smaller or mid-sized desktop monitor. But the sidebar has far less impact on larger desktops or in our mobile layout (when it does become a header button). I personally like having the sidebar open when browsing data... especially for secondary info on detail pages.

In the end, we have a LOT of ideas for new sidebar components, and **hoping** that we can find a place for them in the header is too risky. And this is made even worse when considering that we'll open this up to extensions... then we _need_ a way to support X number of options (while keeping them visible and easily available). Yes, we could probably squeeze all of what we have now into the header, but this change was made to support a more flexible future.

One feature I would like to add (as a header button) is a full-screen mode. This would always be available and would expand your page to cover the viewport (removing the module bar, nav bar, and sidebar). Perhaps this would alleviate some of your frustrations.

Happy to look at mocks, if you want to give them a whirl (I can provide current Figma files)... but I should be upfront that this would be a _huge_ change and isn't something that's likely to happen soon given our dev resources (focused on stability for a while now).

**Comment by oreilles** [comment]
I mean, most displays has more space horizontally than vertically. I'd say at least 1.6 times for most laptops, but often much more for desktop monitor.

So if you have many features to come to the screen it would actually makes more sense to add them where there is the the more space. Also, when browsing tabular data, horizontal space is much more valuable than vertical space because you don't want to keep scrolling to reach the columns you need. That is why excel and any data visualisation tool I have in mind have their display and filter options on the header and none on the sides. 

I think your feeling that the header were being crowded  is largely due to the item count (it takes as much space as all the other stuff for large table) when it could be at the bottom with the the page and limit query options - and also because the buttons and fonts are a bit too large. 😇

Comparing for example Notion's header bar and Directus's, I have the same space left in the middle, when Notion actually includes the quick search, new item button, the layout mode, the layout options, advanced filtering, and even more within the three dots, and that is with explicitly naming the button labels.

<img width="1680" alt="Capture d’écran 2021-01-15 à 02 27 42" src="https://user-images.githubusercontent.com/33065839/104672966-0bc4a880-56e1-11eb-9e38-6ecc9b360615.png">
<img width="1680" alt="Capture d’écran 2021-01-15 à 02 29 14" src="https://user-images.githubusercontent.com/33065839/104672967-0c5d3f00-56e1-11eb-8293-5cb58e7d8d71.png">

**Reply by benhaynes** [reply]
> I mean, most displays has more space horizontally than vertically. I'd say at least 1.6 times for most laptops, but often much more for desktop monitor.

Hilarious, I thought you were arguing my point for a moment! 😉 

>I think your feeling that the header were being crowded is largely due to the item count (it takes as much space as all the other stuff for large table) when it could be at the bottom with the the page and limit query options

If we move item count to the bottom of the page then it needs to be _fixed_, otherwise it'll never be seen. That will decrease the visible page real-estate by blocking the bottom of layouts. I think a better option would be to format that string better. Just making it `1-25 of 1.4M` is a lot shorter!

> and also because the buttons and fonts are a bit too large.

Haha, a bit subjective! They are noticeable, easily clickable/tappable, nicely weighted, language agnostic (no text), and color coded. IMO, better than a few all gray text links. We're trying to avoid text buttons whenever possible since the translations can be WAY longer or shorter... which leads to other issues.

**Reply by oreilles** [reply]
> Just making it 1-25 of 1.4M is a lot shorter!

That's a great idea!

I have an idea that might suit you. For the cards layout, there is second header bar added below the main one. It's pretty empty though. Wouldn't it be able to fit anything and even more than what's currently on the sidebar, with direct acess, without stealing any horizontal space to your layout, and without making your eyes having to adjust to the width change you happen like me to open and close the sidebar pretty frequently ?  It could even be scrollable so there would not even be a limit to what you would want to put in. That would be awesome.

**Reply by benhaynes** [reply]
I'll definitely look into an abbreviated item count notice, seems like an easy win. Also, adding the "overlay mode" so that the sidebar can go over the page instead of pushing it in would solve your issues with _reflow_, since I doubt the 64px on the right (when collapsed) is a big problem for you.

As for the secondary subheader, that's an idea! We can look into how that would look/work in mockups. Still, that's not very modular for extensions, takes up more vertical room on the page, and all "content" of those components would need to open in a drawer/modal. So it's not a silver bullet.

Again, we can research all of this, but what we have now (right info sidebar) has never been reported as an issue by anyone else. _Removing_ that bar would require a significant amount of thought (Notifications, Page Descriptions, Commenting, Settings, etc) and it's just not something we have time to start right now with such a small team.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
